File: carbon/src/main/java/carbon/animation/AnimatedColorStateList.java
package carbon.animation;

import android.animation.ValueAnimator;
import android.content.res.ColorStateList;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.StateSet;
import android.view.animation.AccelerateDecelerateInterpolator;

import java.lang.reflect.Field;
import java.util.Arrays;

import carbon.internal.ArgbEvaluator;

public class AnimatedColorStateList extends ColorStateList {
    private final int[][] states;
    private int[] currentState = null;
    private ValueAnimator colorAnimation = null;
    private int animatedColor;

    private static Field mStateSpecsField, mColorsField, mDefaultColorField;

    static {
        try {
            mStateSpecsField = ColorStateList.class.getDeclaredField("mStateSpecs");
            mStateSpecsField.setAccessible(true);
            mColorsField = ColorStateList.class.getDeclaredField("mColors");
            mColorsField.setAccessible(true);
            mDefaultColorField = ColorStateList.class.getDeclaredField("mDefaultColor");
            mDefaultColorField.setAccessible(true);
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        }
    }

    public static AnimatedColorStateList fromList(ColorStateList list, ValueAnimator.AnimatorUpdateListener listener) {
        int[][] mStateSpecs; // must be parallel to mColors
        int[] mColors;      // must be parallel to mStateSpecs
        int mDefaultColor;

        try {
            mStateSpecs = (int[][]) mStateSpecsField.get(list);
            mColors = (int[]) mColorsField.get(list);
            mDefaultColor = (int) mDefaultColorField.get(list);
            AnimatedColorStateList animatedColorStateList = new AnimatedColorStateList(mStateSpecs, mColors, listener);
            mDefaultColorField.set(animatedColorStateList, mDefaultColor);
            return animatedColorStateList;
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }

        return null;
    }

    public AnimatedColorStateList(int[][] states, int[] colors, final ValueAnimator.AnimatorUpdateListener listener) {
        super(states, colors);
        this.states = states;
        colorAnimation = ValueAnimator.ofInt(0, 0);
        colorAnimation.setEvaluator(new ArgbEvaluator());
        colorAnimation.setDuration(200);
        colorAnimation.setInterpolator(new AccelerateDecelerateInterpolator());
        colorAnimation.addUpdateListener(animation -> {
            synchronized (AnimatedColorStateList.this) {
                animatedColor = (int) animation.getAnimatedValue();
                listener.onAnimationUpdate(animation);
            }
        });
    }

    @Override
    public int getColorForState(int[] stateSet, int defaultColor) {
        synchronized (AnimatedColorStateList.this) {
            if (Arrays.equals(stateSet, currentState)) {
                return animatedColor;
            }
        }
        return super.getColorForState(stateSet, defaultColor);
    }

    public void setState(int[] newState) {
        if (Arrays.equals(newState, currentState))
            return;
        if (currentState != null)
            cancel();

        for (final int[] state : states) {
            if (StateSet.stateSetMatches(state, newState)) {
                int firstColor = super.getColorForState(currentState, getDefaultColor());
                int secondColor = super.getColorForState(newState, getDefaultColor());
                colorAnimation.setIntValues(firstColor, secondColor);
                currentState = newState;
                animatedColor = firstColor;
                colorAnimation.start();
                return;
            }
        }

        currentState = newState;
    }

    private void cancel() {
        colorAnimation.cancel();
    }

    public void jumpToCurrentState() {
        colorAnimation.end();
    }

    public static final Parcelable.Creator<AnimatedColorStateList> CREATOR =
            new Parcelable.Creator<AnimatedColorStateList>() {
                @Override
                public AnimatedColorStateList[] newArray(int size) {
                    return new AnimatedColorStateList[size];
                }

                @Override
                public AnimatedColorStateList createFromParcel(Parcel source) {
                    final int N = source.readInt();
                    final int[][] stateSpecs = new int[N][];
                    for (int i = 0; i < N; i++) {
                        stateSpecs[i] = source.createIntArray();
                    }
                    final int[] colors = source.createIntArray();
                    return AnimatedColorStateList.fromList(new ColorStateList(stateSpecs, colors), null);
                }
            };
}

File: carbon/src/main/java/carbon/animation/AnimatedColorStateList.kt
package carbon.animation

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.content.res.ColorStateList
import android.os.Parcel
import android.os.Parcelable
import android.util.StateSet
import android.view.animation.AccelerateDecelerateInterpolator
import carbon.internal.ArgbEvaluator
import java.lang.reflect.Field
import java.util.*

class AnimatedColorStateList(private val states: Array<IntArray>, colors: IntArray?, listener: AnimatorUpdateListener?) : ColorStateList(states, colors) {
    private var currentState = IntArray(0)
    private val colorAnimation: ValueAnimator
    private var animatedColor = 0

    companion object {
        private var mStateSpecsField: Field? = null
        private var mColorsField: Field? = null
        private var mDefaultColorField: Field? = null
        @JvmStatic
        fun fromList(list: ColorStateList?, listener: AnimatorUpdateListener?): AnimatedColorStateList? {
            val mStateSpecs: Array<IntArray> // must be parallel to mColors
            val mColors: IntArray // must be parallel to mStateSpecs
            val mDefaultColor: Int
            try {
                mStateSpecs = mStateSpecsField!![list] as Array<IntArray>
                mColors = mColorsField!![list] as IntArray
                mDefaultColor = mDefaultColorField!![list] as Int
                val animatedColorStateList = AnimatedColorStateList(mStateSpecs, mColors, listener)
                mDefaultColorField!![animatedColorStateList] = mDefaultColor
                return animatedColorStateList
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
            return null
        }

        val CREATOR: Parcelable.Creator<AnimatedColorStateList?> = object : Parcelable.Creator<AnimatedColorStateList?> {
            override fun newArray(size: Int): Array<AnimatedColorStateList?> {
                return arrayOfNulls(size)
            }

            override fun createFromParcel(source: Parcel): AnimatedColorStateList? {
                val N = source.readInt()
                val stateSpecs = arrayOfNulls<IntArray>(N)
                for (i in 0 until N) {
                    stateSpecs[i] = source.createIntArray()
                }
                val colors = source.createIntArray()
                return fromList(ColorStateList(stateSpecs, colors), null)
            }
        }

        init {
            try {
                mStateSpecsField = ColorStateList::class.java.getDeclaredField("mStateSpecs")
                mStateSpecsField!!.isAccessible = true
                mColorsField = ColorStateList::class.java.getDeclaredField("mColors")
                mColorsField!!.isAccessible = true
                mDefaultColorField = ColorStateList::class.java.getDeclaredField("mDefaultColor")
                mDefaultColorField!!.isAccessible = true
            } catch (e: NoSuchFieldException) {
                e.printStackTrace()
            }
        }
    }

    override fun getColorForState(stateSet: IntArray?, defaultColor: Int): Int {
        synchronized(this@AnimatedColorStateList) {
            if (Arrays.equals(stateSet, currentState) && colorAnimation.isRunning) {
                return animatedColor
            }
        }
        return super.getColorForState(stateSet, defaultColor)
    }

    fun setState(newState: IntArray) {
        synchronized(this@AnimatedColorStateList) {
            if (Arrays.equals(newState, currentState)) return
            colorAnimation.end()
            if (currentState.size != 0) {
                for (state in states) {
                    if (StateSet.stateSetMatches(state, newState)) {
                        val firstColor = getColorForState(currentState, defaultColor)
                        val secondColor = super.getColorForState(newState, defaultColor)
                        colorAnimation.setIntValues(firstColor, secondColor)
                        currentState = newState
                        animatedColor = firstColor
                        colorAnimation.start()
                        return
                    }
                }
            }
            currentState = newState
        }
    }

    fun jumpToCurrentState() {
        colorAnimation.end()
    }

    init {
        colorAnimation = ValueAnimator.ofInt(0, 0)
        colorAnimation.setEvaluator(ArgbEvaluator())
        colorAnimation.duration = 200
        colorAnimation.interpolator = AccelerateDecelerateInterpolator()
        colorAnimation.addUpdateListener { animation: ValueAnimator ->
            synchronized(this@AnimatedColorStateList) {
                animatedColor = animation.animatedValue as Int
                listener!!.onAnimationUpdate(animation)
            }
        }
        colorAnimation.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                animatedColor = colorAnimation.animatedValue as Int
                listener!!.onAnimationUpdate(colorAnimation)
            }
        })
    }
}